compiler/next/include/chpl/uast/Function.h

```diff
@@ -0,0 +1,328 @@
+      formalsChildNum_(formalsChildNum),
+      thisFormalChildNum_(thisFormalChildNum),
+      numFormals_(numFormals),
+      returnTypeChildNum_(returnTypeChildNum),
+      whereChildNum_(whereChildNum),
+      lifetimeChildNum_(lifetimeChildNum),
+      numLifetimeParts_(numLifetimeParts),
+      bodyChildNum_(bodyChildNum),
+      numBodyStmts_(numBodyStmts) {
+
+    assert(-1 <= linkageNameExprChildNum_ &&
+                 linkageNameExprChildNum_ < (ssize_t)children_.size());
+    assert(-1 <= linkageNameExprChildNum_ &&
+                 linkageNameExprChildNum_ < (ssize_t)children_.size());
+    assert(-1 <= formalsChildNum_ &&
+                 formalsChildNum_ < (ssize_t)children_.size());
+    assert(-1 <= thisFormalChildNum_ &&
+                 thisFormalChildNum_ < (ssize_t)children_.size());
+    assert(0 <= numFormals_ &&
+                numFormals_ <= (ssize_t)children_.size());
+    assert(-1 <= returnTypeChildNum_ &&
+                 returnTypeChildNum_ < (ssize_t)children_.size());
+    assert(-1 <= whereChildNum_ &&
+                 whereChildNum_ < (ssize_t)children_.size());
+    assert(-1 <= lifetimeChildNum_ &&
+                 lifetimeChildNum_ < (ssize_t)children_.size());
+    assert(0 <= numLifetimeParts_ &&
+                numLifetimeParts_ <= (ssize_t)children_.size());
+    assert(-1 <= bodyChildNum_ &&
+                 bodyChildNum_ < (ssize_t)children_.size());
+    assert(0 <= numBodyStmts_ &&
+                numBodyStmts_ <= (ssize_t)children_.size());
+    assert(isExpressionASTList(children_));
+  }
+  bool contentsMatchInner(const ASTNode* other) const override;
+  void markUniqueStringsInner(Context* context) const override;
+
+ public:
+  ~Function() override = default;
+
+  Linkage linkage() const { return this->linkage_; }
+  Kind kind() const { return this->kind_; }
+  ReturnIntent returnIntent() const { return this->returnIntent_; }
+  bool isInline() const { return this->inline_; }
+  bool isOverride() const { return this->override_; }
+  bool throws() const { return this->throws_; }
+
+  /**
+   Return the linkage name expression, e.g. "f_c_name"
+   in the below, or nullptr if there is none.
+ 
+   \rst
+    .. code-block:: chapel
+
+        extern "f_c_name" proc f(arg) { }
+    \endrst
+   */
+  const Expression* linkageNameExpression() const {
+    if (linkageNameExprChildNum_ >= 0) {
+      const ASTNode* ast = child(linkageNameExprChildNum_);
+      assert(ast->isExpression());
+      return (const Expression*) ast;
+    } else {
+      return nullptr;
+    }
+  }
+
+  // TODO: should formals() result in Formal or FormalDecl?
+  //       should getFormal() result in the same type?
+  //       do we need two variants of these?
+
+  /**
+   Return a way to iterate over the formals, including the method
+   receiver, if present, as the first formal.
+   */
+  ASTListIteratorPair<FormalDecl> formals() const {
+    if (numFormals_ == 0) {
```

Comment:
It might be good to use the numFormals() method instead of the underlying value - we'll better avoid code drift if we use the same interface we present to outside the class